### DISCUSSION TITLE
Data duplication problem when I use @apollo/client with nextjs on route change?

### DISCUSSION THREAD

**Discussion by siamahnaf** [vercel/next.js#46823] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

I am building a project using nextjs and @apollo/client. Here I am only fetching my data. My api is provider from other source.

Here I am facing one problem. When I change my page and again go to previous page I can I see data get twice or duplicate. How can I solve my problem.

I follow this repo for setup- [https://github.com/vercel/next.js/tree/canary/examples/with-apollo](https://github.com/vercel/next.js/tree/canary/examples/with-apollo)

![Video_2023_03_04-1_edit_1](https://user-images.githubusercontent.com/78312010/222915378-0fe44560-2964-4469-9b67-8f57e458bfd1.gif)

Here is my code-

`apolloClient.ts`-

```
import { useMemo } from "react"
import { ApolloClient, HttpLink, InMemoryCache, from, NormalizedCacheObject } from "@apollo/client"
import { onError } from "@apollo/client/link/error"
import { concatPagination } from "@apollo/client/utilities"
import merge from "deepmerge"
import isEqual from "lodash/isEqual"

export const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__"

let apolloClient: ApolloClient<NormalizedCacheObject>;

const errorLink = onError(({ graphQLErrors, networkError }) => {
    if (graphQLErrors)
        graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
                `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
        )
    if (networkError) console.log(`[Network error]: ${networkError}`)
})

const httpLink = new HttpLink({
    uri: "http://localhost:3001/ecampus",
    credentials: "include"
})

function createApolloClient() {
    return new ApolloClient({
        ssrMode: typeof window === 'undefined',
        link: from([errorLink, httpLink]),
        cache: new InMemoryCache({
            typePolicies: {
                Query: {
                    fields: {
                        allPosts: concatPagination(),
                    },
                },
            },
        }),
        defaultOptions: {
            mutate: {
                errorPolicy: "all"
            }
        }
    })
}

export function initializeApollo(initialState = null) {
    const _apolloClient = apolloClient ?? createApolloClient()
    if (initialState) {
        const existingCache = _apolloClient.extract()
        const data = merge(existingCache, initialState, {
            arrayMerge: (destinationArray, sourceArray) => [
                ...sourceArray,
                ...destinationArray.filter((d) =>
                    sourceArray.every((s) => !isEqual(d, s))
                ),
            ],
        })
        _apolloClient.cache.restore(data)
    }
    if (typeof window === 'undefined') return _apolloClient
    if (!apolloClient) apolloClient = _apolloClient

    return _apolloClient
}

export function addApolloState(client: any, pageProps: any) {
    if (pageProps?.props) {
        pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract()
    }

    return pageProps
}

export function useApollo(pageProps: any) {
    const state = pageProps[APOLLO_STATE_PROP_NAME]
    const store = useMemo(() => initializeApollo(state), [state])
    return store
}
```

And in `_app.tsx`-

```
export default function App({ Component, pageProps }: AppProps) {
  const apolloClient = useApollo(pageProps);
  return (
    <ApolloProvider client={apolloClient}>
      <main className={`${inter.variable} font-sans`}>
        <Component {...pageProps} />
      </main>
    </ApolloProvider>
  )
}
```

And then I one page named `section.tsx`-

```
export const getServerSideProps: GetServerSideProps = async (ctx) => {
    const apolloClient = initializeApollo()
    await apolloClient.query({
        query: GET_SECTION_LIST,
        variables: { searchInput: {} },
        context: {
            headers: {
                cookie: ctx.req.headers.cookie,
                "user-agent": ctx.req.headers["user-agent"],
                "x-forwarded-for": ctx.req.socket.remoteAddress
            }
        }
    })
    return addApolloState(apolloClient, {
        props: {},
    })
}
```

After fetching data from `getServerSideProps`, I use data in one component name is `list.tsx`. `list.tsx` called in `section.tsx` page-

`const { data, error } = useQuery<GetSectionListData>(GET_SECTION_LIST, { variables: { searchInput: {} } });`

I use this data in a table that you can see on the gif.

Here is my packages-

```
"@apollo/client": "^3.7.9"
"deepmerge": "^4.3.0"
"lodash": "^4.17.21"
"graphql": "^16.6.0",
```

How can I solve my problem? Please help me!

### Which area(s) of Next.js are affected? (leave empty if unsure)

Data fetching (gS(S)P, getInitialProps)

### Link to the code that reproduces this issue

https://github.com/siamahanf/campus

### To Reproduce

I have no idea

### Describe the Bug

I have no idea

### Expected Behavior

It should work perfectly!

**Comment by siamahnaf** [comment]
Anyone can help me?

**Comment by icyJoseph** [comment]
@siamahanf Could you debug this part?

```js
    const data = merge(existingCache, initialState, {
            arrayMerge: (destinationArray, sourceArray) => [
                ...sourceArray,
                ...destinationArray.filter((d) =>
                    sourceArray.every((s) => !isEqual(d, s))
                ),
            ],
        })
```

Like so perhaps, console.log is not the best, but just an idea:

```js
    const data = merge(existingCache, initialState, {
            arrayMerge: (destinationArray, sourceArray) => {
             let ret = [
                ...sourceArray,
                ...destinationArray.filter((d) =>
                    sourceArray.every((s) => !isEqual(d, s))
                ),
            ];
            console.log("RET", ret);
            return ret;
           },
        })
```

Can you confirm that it removes duplicates?

**Reply by icyJoseph** [reply]
@siamahanf, alright then do these items have a unique identifier? 

If they do, then you can do a pattern like:

```js
// pseudo-code
const seen = new Set()
const uniqueById = items.reduce((prev, curr) => {
  if (seen.has(curr.id)) return prev

  seen.add(curr.id)
  prev.push(curr)
  return prev
}, [])
```

Of course this assumes that we want to keep `items`-by-id as they are placed from `left-to-right` on the list, otherwise `items.reduceRight` is another option, but still both assuming order is the key element.

The interesting thing is why do we see different objects, perhaps there's a timestamp attached to them?

**Reply by siamahnaf** [reply]
Where I write it?

**Reply by icyJoseph** [reply]
That'd be where you do the arrayMerge I guess, it's been years since I used apollo, but I have a feeling that's where the issue happens, and you seem to confirm it in the other message:

```ts
const data = merge(existingCache, initialState, {
  arrayMerge: (destinationArray, sourceArray) => {
    let merged = [
      ...sourceArray,
      ...destinationArray.filter((d) =>
        sourceArray.every((s) => !isEqual(d, s))
      ),
    ];
    const seen = new Set();
    const uniqueById = merged.reduce((prev, curr) => {
      if (!seen.has(curr.id)) {
        seen.add(curr.id);
        prev.push(curr);
      }
      return prev;
    }, []);
    return merged;
  },
});
```

Something like that, assuming of course that first in line matters, and that there's an `id` and ignoring why there'd be attempts at duplicating it anyway... or in the first place.

**Reply by siamahnaf** [reply]
I am getting this error-

![image](https://user-images.githubusercontent.com/78312010/223320326-f54a2433-d618-430d-831f-15035d29b866.png)

**Reply by icyJoseph** [reply]
Reduce may take a second parameter, in this case you should pass an empty array. `reduce(callback, [])` - Fixed snippet above, but the pseudo code I posted initially had it.

Regardless, I think this is but a patch, and not a real fix, if it works at all.

**Reply by siamahnaf** [reply]
Okay, successfully solve the data duplication problem. But I am facing new problem with cache related. I already place a issue on this apollo client. Can you help me, please with that issue, please, please..

Here is link- https://github.com/apollographql/apollo-client/issues/10627

**Reply by icyJoseph** [reply]
Sorry, gonna be short on time today.

It looks like you should be able to debug that one though.

Just for sake of testing, if you replace reduce with reduceRight, does it work?

You need to make sure that the cache updating logic receives the inputs you expect.

Don't forget to mark as answered!

**Comment by siamahnaf** [comment]
Not, it didn't removes the duplicate.
